File: Codility/Lesson2 - Arrays/CyclicRotation.swift
/*
    Codility Lesson 2 - Arrays

    CyclicRotation
    : Rotate an array to the right by a given number of steps.
  
    - Level of Difficulty: Painless
    - Task Score: 100%
    - Correntness: 100%  
*/

import Foundation
import Glibc

// you can write to stdout for debugging purposes, e.g.
// print("this is a debug message")

public func solution(_ A : inout [Int], _ K : Int) -> [Int] {
    guard !A.isEmpty else { return A }
    guard K > 0 else { return A }
    guard K % A.count != 0 else { return A }
    
    let subrange = (A.endIndex - (K % A.count))..<A.endIndex
    let elementsToShift = A[subrange]
    A.removeSubrange(subrange)
    A.insert(contentsOf: elementsToShift, at: 0)
    
    return A
}

File: Codility/CyclicRotation.java
/*
    CyclicRotation
    : Rotate an array to the right by a given number of steps.

    - Difficulty Level: PAINLESS

    - Task Score: 100%
    - Correctness: 100%
 */

import java.util.*;

class Solution {
    public int[] solution(int[] A, int K) {
        if (A.length == 0) { return A; }

        int length = A.length;
        int rotations = K % length;

        if (rotations == 0) {
            return A;
        } else {
            int[] rotated = new int[length];

            int[] latter = Arrays.copyOfRange(A, (length - rotations), length);
            int[] former = Arrays.copyOfRange(A, 0, (length - rotations));

            for (int i = 0; i < latter.length; i++) {
                rotated[i] = latter[i];
            }

            for (int i = 0; i < former.length; i++) {
                rotated[i + latter.length] = former[i];
            }

            return rotated;
        }
    }
}